pkg/compactor/blocks_cleaner.go

```diff
@@ -402,17 +411,28 @@ func (c *BlocksCleaner) deleteBlocksMarkedForDeletion(ctx context.Context, idx *
 
 // cleanUserPartialBlocks delete partial blocks which are safe to be deleted. The provided partials map
 // and index are updated accordingly.
-func (c *BlocksCleaner) cleanUserPartialBlocks(ctx context.Context, partials map[ulid.ULID]error, idx *bucketindex.Index, userBucket objstore.InstrumentedBucket, userLogger log.Logger) {
+func (c *BlocksCleaner) cleanUserPartialBlocks(ctx context.Context, partials []bucketindex.PartialBlock, idx *bucketindex.Index, delay time.Duration, userBucket objstore.InstrumentedBucket, userLogger log.Logger) {
 	// Collect all blocks with missing meta.json into buffered channel.
 	blocks := make([]ulid.ULID, 0, len(partials))
 
-	for blockID, blockErr := range partials {
+	for _, p := range partials {
 		// We can safely delete only blocks which are partial because the meta.json is missing.
-		if !errors.Is(blockErr, bucketindex.ErrBlockMetaNotFound) {
+		if !errors.Is(p.Reason, bucketindex.ErrBlockMetaNotFound) {
 			continue
 		}
-
-		blocks = append(blocks, blockID)
+		// Check if partial block is older than delay period, and mark for deletion if necessary
+		if delay > 0 {
+			maxTime := time.Unix(p.MaxTime/1000, 0)
```

Comment:
Is there a reason not to use `time.UnixMilli` for simplicity?
```suggestion
			maxTime := time.UnixMilli(p.MaxTime)
```

Reply:
`time.UnixMilli` is go 1.17+. Not sure if we want to increase the minimum required version?

Reply:
I changed this in the latest commit but will revert if we want to be compatible with 1.16